### DISCUSSION TITLE
Rules, moisture sensor and rain

### DISCUSSION THREAD

**Discussion by icho40** [arendst/Tasmota#22118] [Q&A]
Hi,
I want a soil moisture sensor that switches on a water pump when the soil is dry, 
but only when it is not raining.
As a beginner, I'm trying out rules for the first time.

I configured my ESP32 as follows:

Switch1 on unused GPIO
Relay1 on unused GPIO
Humidity sensor on analog pin:
ADC range
ANALOG#Range1
Range: Minimum 49 (completely wet) Maximum 80 (dry)
Pump defined as Relay2

Rain sensor as a digital sensor: 
Switch5 
Switchmode5 2

SetOption114 1

This works well:
Rule1 ON ANALOG#Range1 >60 DO Power2 ON ENDON ON ANALOG#Range1 <50 DO Power2 OFF ENDON 
Rule1 1

But when I try to include the rain sensor:
Rule1 ON ANALOG#Range1  >60 IF (Switch5#Action = OFF)  DO Power2 ON ENDIF ENDON ON ANALOG#Range1  <50 DO Power2 OFF  ENDON 
Rule1 1

I see in the console:
 stat/tasmota_4E4E64/RESULT = {"Switch5":{"Action":"ON"}}
but the Power2 (relay2) is still on.

Can anyone give me a hint as to where my mistake might be?

**Comment by sfromis** [comment]
As rules are event-based, the trigger expressions with `#` can **only** be used in the trigger part of rules, between `on` and `do`, where only a single condition is possible. What you can do is to use a [rule variable](https://tasmota.github.io/docs/Rules/#rule-variables) like `%switch2%` in an `if` condition.

Note that you should not try to assign the gpio dropdown skipping numbers, but always start the indexes from 1 and counting upwards. They are not mapping numbers, but only to indicate the sequence, hence the starting point of 1 without skipped numbers.

More generally, when you have ESP32, I can recommend the [Berry](https://tasmota.github.io/docs/Berry/) programming language, which is a bit like a simplified Python, where you can use more normal programming constructs. Obviously still a learning curve, but maybe a better one to take than the one to get up to speed with the arcane and "less helpful" rules, which will just fail silently if you do not get all the details just right.

**Reply by icho40** [reply]
Unfortunately, I'm not a programmer, 
just a retired court bailiff from germany.
I've been using Linux for years, 
but I can't do more than a few simple shell scripts.
I tried to learn C programming a while ago, 
but when it started with pointers, I gave up

I've now tried Berry scripting:

def rain(n)
if n==1 return "it rains" end
return "norain"
end

works:
rain(0)
norain
rain(1)
it rains

Unfortunately, this doesn't work:
def rain(n)
if n==1 return tasmota.set_power(0,1) end
return "norain"
end

rain(0)
norain
rain(1)
stack traceback:
<native>: in native function
<unknown source>: in function `rain`
<unknown source>: in function `main`
Feedback geben


I think I'll just add another relay that will shut off 
the other relays when it rains.

**Reply by sfromis** [reply]
You can see more details (like `type_error`) if you watch the regular console (instead of the Berry one), but the issue is that `tasmota.set_power` takes a boolean (`true` / `false`) as the 2nd argument, not a number.

BTW. While there is nothing "wrong" about returning the result from `tasmota.set_power` from your `rain()` function, the "regular" way would be a bit longer, like:
```
def rain(n)
  if n==1 
    tasmota.set_power(0, true)
    return 'rain'
  end
  return 'norain'
end
```

For illustrative purposes, you could also code like this:
```
def rain(n)
  tasmota.set_power(0, n==1)       # comparison resulting in true/false, powers off if "norain"
  return n==1 ? 'rain' : 'norain'  # ternary operator, like if/else but shorter
end
```
(Not really meant as a suggestion for you to follow, as it does not lend itself to expansion)
